### PR TITLE
feature(gatsby): Add CLI for managing flags

### PR DESCRIPTION
**Pull request by KyleAMathews**
Follow-up to https://github.com/gatsbyjs/gatsby/pull/28296

- adds cli command `gatsby flags` to manage which flags are enabled
- adds resource to create/destroy flags in gatsby-config.js

TODOs
- [ ] catch babel parse errors
- [ ] move to enquirer from Ink
- [ ] add support for `--enable` & `--disable` flags

### CODE REVIEW

File: packages/gatsby-dev-cli/src/utils/check-deps-changes.js [hunk 1]
```diff
@@ -61,15 +61,15 @@ exports.checkDepsChanges = async ({
...
+          .then((error, response) => {
             if (response && response.statusCode === 200) {
               return resolve(JSON.parse(response.body))
             }
-
-            return reject(error)
-          }
-        )
+          })
+          .catch(e => {
```
Comment (KyleAMathews) [inline packages/gatsby-dev-cli/src/utils/check-deps-changes.js:70]: @pieh this was failing on the new package

File: packages/gatsby/src/experiments.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
+const activeExperiments = [
+  {
+    name: `FAST_DEV`,
```
Comment (KyleAMathews) [inline packages/gatsby/src/experiments.ts:3]: The initial set of experiments is from the dev speedup work — we'll add/remove from here as we add/graduate/pull out experiments

File: packages/gatsby/src/services/initialize.ts [hunk 3]
```diff
@@ -178,6 +178,71 @@ export async function initialize({
...
+      addIncluded(experiment)
+    })
+
+    validConfigExperiments = _.uniq(validConfigExperiments)
+
+    // TODO remove experiments that longer exist.
+
+    //  set process.env for remaining
+    validConfigExperiments.forEach(experiment => {
+      process.env[`GATSBY_EXPERIMENTAL_${experiment.name}`] = `true`
```
Comment (KyleAMathews) [inline packages/gatsby/src/services/initialize.ts:218]: add to process.env any active experiments

File: packages/gatsby-recipes/src/providers/gatsby/experiment.js [hunk 4]
```diff
@@ -0,0 +1,301 @@
+import fs from "fs-extra"
```
Comment (KyleAMathews) [inline packages/gatsby-recipes/src/providers/gatsby/experiment.js:1]: resource for manipulating experiments in gatsby-config.js

File: packages/gatsby-experiments-cli/src/index.js [hunk 5]
```diff
@@ -0,0 +1,108 @@
+const React = require("react")
```
Comment (KyleAMathews) [inline packages/gatsby-experiments-cli/src/index.js:1]: little Ink app for managing experiments

**Review by wardpeet — CHANGES_REQUESTED** [review]
Can we scope this PR down to only enable experiments through gatsby-config.js. Adding a new package with its own ink, ... is not great. It makes gatsby-cli even bigger. As a minimum we can show a list of experiments and some descriptions and tell them to enable it in the config.

When create-gatsby is part of gatsby new we can piggyback on their usage of enquirer instead. I'm also worried about using recipes (experimental tech) for this feature.

File: packages/gatsby/src/services/initialize.ts [hunk 6]
```diff
@@ -178,6 +178,71 @@ export async function initialize({
     })
   }
 
+  // Setup experiments
+  if (config && config.__experiments) {
+    const experiments = require(`../experiments`).default
+    const configExperiments = config.__experiments
```
Comment (wardpeet) [inline packages/gatsby/src/services/initialize.ts:184]: shouldn't we just go for experiments? __ might be hard for people to discover.

```suggestion
    const configExperiments = config.experiments
```

Comment (KyleAMathews) [reply packages/gatsby/src/services/initialize.ts:184]: I was going off your original proposal 🤣

But yeah, no reason to add the underscores.

**Comment by KyleAMathews** [comment]
>Adding a new package with its own ink, ... is not great

yeah — if the main CLI were on Ink v3 already this wouldn't be much more but as it is (just checked) it's an extra 1.5m so that's no good. It'd be easy to reimplement w/ enquirer so I'll just do that.

>I'm also worried about using recipes (experimental tech) for this feature.

We need to distinguish when discussing recipes between its two parts — part 1 is a read/write API for Gatsby resources and part 2 is a runtime for install scripts written in MDX built on top of part 1.

The read/write API is straightforward tech and I wouldn't call it experimental. A new runtime is a bit more out there :-)

`create-gatsby` is already relying on the recipes API for installing packages/plugins and there's a number of future use cases for using it in Admin.

**Comment by KyleAMathews** [comment]
And happy to split the PR into two if that makes reviewing easier but a CLI management tool feels essential as otherwise there's no where to teach people about the experiments and it reduces friction significantly for trying things.

File: packages/gatsby-dev-cli/src/utils/check-deps-changes.js [hunk 7]
```diff
@@ -61,15 +61,13 @@ exports.checkDepsChanges = async ({
...
             if (response && response.statusCode === 200) {
-              return resolve(JSON.parse(response.body))
+              resolve(JSON.parse(response.body))
             }
-
-            return reject(error)
-          }
-        )
+          })
+          .catch(e => reject(e))
```
Comment (pvdz) [inline packages/gatsby-dev-cli/src/utils/check-deps-changes.js:70]: Why not await?

```js
const [error, response] = got(`https://unpkg.com/${packageName}/package.json`)
let localPKGjson
if (error) { // Assuming got doesn't actually throw... I'm just unwrapping the `.then`
  throw new Error(error)
}
if (response && response.statusCode === 200) {
  localPKGjson = JSON.parse(response.body)
}
```

**Comment by pieh** [comment]
I propose to use object notation:
```
experiments: {
  DEV_SSR: true
}
```

and not array, to not limit ourselves in case something will require configuration in the future and not just enabled/disabled toggle

**Comment by wardpeet** [comment]
I wouldn't add cli for first pass, teaching can be done by docs and telling people how to enable it. No big deal, CLI seems too much as a starting point. Things can go wrong with babel transforms when people are using gatsby-config differently than we expect, we seen this with recipes.

keep it simple, only add experiments to config, nothing else.

**Comment by KyleAMathews** [comment]
>Things can go wrong with babel transforms when people are using gatsby-config differently than we expect, we seen this with recipes.

This is easy to work around — babel throws an error when it can't parse something and the CLI catches it and says "we weren't able to add the feature flag to your gatsby-config.js — please add it manually like this" etc.

**Comment by KyleAMathews** [comment]
But yeah, I'll do this on a subsequent PR.

**Comment by KyleAMathews** [comment]
With the CLI, we can also add a one-liner command `gatsby feature --enable FAST_DEV` and `gatsby feature --disable FAST_DEV`

**Comment by ascorbic** [comment]
I'd echo those saying that we should scope this down to just using the gatsby-config, without the cli, and I say that as a fan of interactive CLIs! I think anybody who is confident enough to be enabling experiments can do it in the config. Adding a whole new cli and recipe provider is a bit much.

File: packages/gatsby/src/services/initialize.ts [hunk 8]
```diff
@@ -178,6 +178,71 @@ export async function initialize({
     })
   }
 
+  // Setup experiments
+  if (config && config.__experiments) {
+    const experiments = require(`../experiments`).default
```
Comment (ascorbic) [inline packages/gatsby/src/services/initialize.ts:183]: Can this be a regular ESM import up at the top?

**Comment by KyleAMathews** [comment]
@ascorbic I did pull this out into its own PR (which got merged yesterday) — https://github.com/gatsbyjs/gatsby/pull/28296

Did you think the cli was "too much" for the initial PR or in general? The two compelling use cases for me a) we can tell people to just run `gatsby flags --enable FAST_DEV` to add something as that's still far less friction to get someone to give a feature a try and report back and b) being able to see a list of flags as otherwise, there's no way for people to discover new features unless they're an avid reader of the release notes.

It was only a few hours to build the resource/CLI so I'm fine either way.

**Comment by ascorbic** [comment]
@KyleAMathews A bit much in general. I'm super keen on using the CLI to make stuff simpler, but this isn't somehting that most people will be using, and the people who do will be the sort who are fine to edit the config. if there was a generalisable way of editing the config without needing to create a new babel parser and everything then I might lean more towards it, but this is not going to be something that most users will ever need. Perhaps a way to aid discoverability as well as making things easier to debug would be to add a section to the end of the `gatsby info` output that lists all experiments and whether or not they're enabled?

**Comment by KyleAMathews** [comment]
I like writing custom babel parsers but maybe I'm weird :-D — much of the code there is already abstracted and shared with the site-metadata & plugin resources & there's some more repetitive code there that could be pulled out. It's essentially the same thing as the site-metadata resource.

Yeah I was thinking that when someone has enabled a flag that we show them other flags underneath that they can enable as well. That's perhaps enough as most people I agree will never need/want this.

**Comment by ascorbic** [comment]
Haha. So do I. There seem to be a lot of us in the company! My issue with it is that it's fragile, and there's a lot of repetition and it's hard to reason about. tbh, a general interface to gatsby-config that didn't need every new key to have a whole parser written would be nice addition to core. An AST node visitor isn't the most intuitive interface to setting configuration values.

**Comment by KyleAMathews** [comment]
Yeah a generalized API would be very handy.

**Comment by KyleAMathews** [comment]
Not essential

**Pull request closed without merging**